sound/soc/sof/ipc4-topology.c

```diff
@@ -1215,13 +1248,41 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		struct snd_sof_widget *pipe_widget;
 		struct sof_ipc4_pipeline *pipeline;
 
-		pipe_widget = swidget->spipe->pipe_widget;
-		pipeline = pipe_widget->private;
 		ipc4_copier = (struct sof_ipc4_copier *)swidget->private;
 		gtw_attr = ipc4_copier->gtw_attr;
 		copier_data = &ipc4_copier->data;
 		available_fmt = &ipc4_copier->available_fmt;
 
+		pipe_widget = swidget->spipe->pipe_widget;
+		pipeline = pipe_widget->private;
+
+		if (pipeline->use_chain_dma) {
+			u32 host_dma_id;
+			u32 fifo_size;
+
+			host_dma_id = platform_params->stream_tag - 1;
+			pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(host_dma_id);
+
+			/* Set SCS bit for S16_LE format only */
+			if (params_format(fe_params) == SNDRV_PCM_FORMAT_S16_LE)
+				pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
+
+			/*
+			 * Despite its name the bitfield 'fifo_size' is used to define DMA buffer
+			 * size. The expression calculates 2ms buffer size.
+			 */
+			fifo_size = DIV_ROUND_UP((SOF_IPC4_CHAIN_DMA_BUF_SIZE_MS *
+						  params_rate(fe_params) *
+						  params_channels(fe_params) *
+						  params_physical_width(fe_params)), 8000);
+			pipeline->msg.extension |= SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE(fifo_size);
+
+			/* Set node_id to invalid value. It shouldn't be used with chained DMA. */
+			copier_data->gtw_cfg.node_id = SOF_IPC4_INVALID_NODE_ID;
```

Comment:
why is this needed? Its not like we ever send the IPC to set up the copier widget for chained DMA, right?

Reply:
IIRR it was @RanderWang 's suggestion. AFAIU the reasoning is to cause an error, if a chain dma swidget is ever used in a wrong place.

Reply:
@jsarha that would be wrong and would lead to memory leaks. We allocate memory based on the node ID durig setup() and if you modify it during prepare, what happens during free would be anybody's guess.

Reply:
Wrong yes, the idea is only to catch such a bug early. Or do you mean that just making the assignment here would cause a leak? That would be surprising as  copier_data->gtw_cfg.node_id is assigned in so many other branches of the function.

Reply:
sorry I got confused. I thpought we allocated memory based on node_id but its really only based on dai_type. So I guess this is fine

Reply:
But I still think it is not right set to invalid here. The idea is to keep the tplg identical for both chained DMA and regular pipelines. So I do not see what error you would catch with this.

Reply:
Here is the original comment from @RanderWang 's: https://github.com/thesofproject/linux/pull/3953#discussion_r1017558014 

I guess we can wait until he wakes up, if there is something more to the suggestion than I  can explain, before making the final decision.

Reply:
Chain-DMA does't support time stamp feature which will check node to skip this case.

Reply:
````c
static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
					    struct snd_pcm_substream *substream,
					    struct snd_sof_pcm_stream *stream,
					    struct sof_ipc4_timestamp_info *time_info)
{
	struct sof_ipc4_copier *host_copier = time_info->host_copier;
	struct sof_ipc4_copier *dai_copier = time_info->dai_copier;
	struct sof_ipc4_pipeline_registers ppl_reg;
	u64 stream_start_position;
	u32 dai_sample_size;
	u32 ch, node_index;
	u32 offset;

	node_index = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
	if (node_index == SOF_IPC4_INVALID_NODE_ID)
		return -EINVAL;
````

Reply:
@jsarha so it looks like we need to set the node ID to invalid for this particular reason that we do no support timestamping feature with chained DMA. Can you please update your comment here so it is clear why we need this?